### PR TITLE
Add Binds, WhiteBinds, Whitelistedrange to CConnman::Options

### PR DESCRIPTION
**Pull request by benma**
Subset of #9897, with the goal to decouple command line arg parsing and make AppInitMain() smaller.

### CODE REVIEW

**Comment by benma** [comment]
Pinging @theuni

File: src/net.cpp [hunk 1]
```diff
@@ -64,6 +64,14 @@
 #endif
 #endif
 
+/** Used to pass flags to the Bind() function */
+enum BindFlags {
```
Comment (practicalswift) [inline src/net.cpp]: Minor nit, but what about switching to C++11 strongly typed enumerations (`enum class`) while we are at it?

Comment (benma) [inline src/net.cpp]: Great idea,  but unfortunately, bitwise operations don't work with those (unless you manually add those functions to the class) :(

See http://blog.bitwigglers.org/using-enum-classes-as-type-safe-bitmasks/

Comment (practicalswift) [inline src/net.cpp]: Ah, of course. Sorry!

**Review by ryanofsky — COMMENTED** [review]
utACK 7313bfec5354f44571929249ee3672b8c38675a3. This seems like an improvement, though I also think it might be a little better if the connman options were vectors of strings, so AppInit would only be dealing with translating arguments (and the loops could be eliminated), and connman would deal with everything network related (lookups).

**Comment by benma** [comment]
@ryanofsky thanks. I actually did it like this in a first iteration, but changed it because I figured that the arguments should evaluated as much as possible before entering ConnMan.

It was a bit weird having the error messages that relate to `-whitelist`, `-bind`, `-whitebind` still in ConnMan as opposed to where the arguments are actually parsed.

**Comment by theuni** [comment]
I believe this is OK. I have a PR coming up which (after lots of thought) moves resolving into its own class, of which CConnman is one user. This way other layers are able to use the resolver without necessarily having to go through CConnman.

So, concept ACK. Reviewing.

File: src/init.cpp [hunk 2]
```diff
@@ -1644,6 +1594,28 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     connOptions.nMaxOutboundTimeframe = nMaxOutboundTimeframe;
     connOptions.nMaxOutboundLimit = nMaxOutboundLimit;
 
+    if (gArgs.IsArgSet("-bind")) {
```
Comment (theuni) [inline src/init.cpp]: This looks like a change in behavior when setting "-bind=foo -listen=0". Granted, that combination makes no sense.

I think we should probably just check for that combination, and exit with an error in that case.

File: src/net.cpp [hunk 3]
```diff
@@ -2183,7 +2191,41 @@ NodeId CConnman::GetNewNodeId()
...
+        if ((flags & BF_REPORT_ERROR) && clientInterface) {
+            clientInterface->ThreadSafeMessageBox(strError, "", CClientUIInterface::MSG_ERROR);
+        }
+        return false;
+    }
+    return true;
+}
+
+bool CConnman::InitBinds(const std::vector<CService>& binds, const std::vector<CService>& whiteBinds) {
+    if (fListen) {
```
Comment (theuni) [inline src/net.cpp]: See other comment. It doesn't make sense to have binds/whitebinds set at all if !fListen.

File: src/net.cpp [hunk 4]
```diff
@@ -2190,7 +2191,41 @@ NodeId CConnman::GetNewNodeId()
...
+        bool fBound = false;
+        for (const auto& addrBind : binds) {
```
Comment (theuni) [inline src/net.cpp]: Mm, this moves the logic away from the caller and into CConnman, which is what we're trying to avoid by adding the heaps of options.

I think I'd prefer to see flags passed in along with CServices, so that CConnman can just do what it's told. So either a tiny struct, or just a pair.

Then, CConnman will try to bind the BF_EXPLICIT addresses, and if it fails, tries to bind the rest.

Also, in that case, the last failed bind wouldn't need to have BF_REPORT_ERROR set, we'd just notice the false return and act on it accordingly.

Sound reasonable?

**Review by theuni — CHANGES_REQUESTED** [review]
Concept ACK. Thanks for working on this!

File: src/net.cpp (hunk 4 above)
Comment (benma) [inline src/net.cpp]: > Mm, this moves the logic away from the caller and into CConnman, which is what we're trying to avoid by adding the heaps of options.
> I think I'd prefer to see flags passed in along with CServices, so that CConnman can just do what it's told. So either a tiny struct, or just a pair.

The logic would be in `CConnman` all the same, just the data structure you pass the info with is changed from two vectors to one vector of pairs, which is equivalent. The code looks more complicated from what I can tell. Or did I misunderstand your request?

I added this in a separate commit to see the diff, I drop it again if needed.

> Then, CConnman will try to bind the BF_EXPLICIT addresses, and if it fails, tries to bind the rest.

At the moment, the rest is only attempted if no explicit binds are specified. If they fail, all fails, which makes sense to me.

> Also, in that case, the last failed bind wouldn't need to have BF_REPORT_ERROR set, we'd just notice the false return and act on it accordingly.

Act how, apart from reporting the error (which this flag does)?

File: src/init.cpp (hunk 2 above)
Comment (benma) [inline src/init.cpp]: Done.

File: src/net.cpp (hunk 3 above)
Comment (benma) [inline src/net.cpp]: Done.

**Comment by sipa** [comment]
@theuni Can you check whether your requested changes are addressed?

File: src/net.cpp (hunk 4 above)
Comment (theuni) [inline src/net.cpp]: >>Mm, this moves the logic away from the caller and into CConnman, which is what we're trying to avoid by adding the heaps of options.
    I think I'd prefer to see flags passed in along with CServices, so that CConnman can just do what it's told. So either a tiny struct, or just a pair.

> The logic would be in CConnman all the same, just the data structure you pass the info with is changed from two vectors to one vector of pairs, which is equivalent. The code looks more complicated from what I can tell. Or did I misunderstand your request?

The logic would still be in CConnman, but the caller would have the ability to dictate exactly how the binds are setup. For example, there's no way to specify "listen on this address as well as the localhost fallbacks".

I think this is ok for now, though. We can add options as things come up.

File: src/net.cpp [hunk 5]
```diff
@@ -2230,7 +2231,38 @@ NodeId CConnman::GetNewNodeId()
...
 }
 
-bool CConnman::Start(CScheduler& scheduler, std::string& strNodeError, Options connOptions)
+
+bool CConnman::Bind(const CService &addr, unsigned int flags) {
+    if (!(flags & BF_EXPLICIT) && IsLimited(addr))
+        return false;
+    std::string strError;
+    if (!BindListenPort(addr, strError, (flags & BF_WHITELIST) != 0)) {
+        if ((flags & BF_REPORT_ERROR) && clientInterface) {
```
Comment (theuni) [inline src/net.cpp]: Well done! Thanks for using the local instance.

File: src/net.cpp [hunk 6]
```diff
@@ -2230,7 +2231,38 @@ NodeId CConnman::GetNewNodeId()
...
+
+bool CConnman::InitBinds(const std::vector<std::pair<CService, Options::BindType> >& binds) {
+    bool fBound = false;
+    for (const auto& addrBind : binds) {
+        unsigned int flags = BF_EXPLICIT | BF_REPORT_ERROR;
+        if (addrBind.second == Options::BindType::WhiteBind)
+            flags |= BF_WHITELIST;
+        fBound |= Bind(addrBind.first, flags);
+    }
+    if (binds.empty()) {
```
Comment (theuni) [inline src/net.cpp]: Looks like there's no way to run without listening now.

Sorry for not being more clear here: https://github.com/bitcoin/bitcoin/pull/10496#discussion_r120199183. I was suggesting that we shouldn't get into InitBinds in the first place if binding isn't allowed.

We'll either need to change this so that its caller uses the ugly global as before:
``` c++
    if(fListen && !InitBinds(...)) {
        ...
    }
```
or add a flag for fallback connections, which InitBinds() only uses if nothing else is already bound.

File: src/net.cpp [hunk 7]
```diff
@@ -2252,11 +2284,19 @@ bool CConnman::Start(CScheduler& scheduler, std::string& strNodeError, Options c
 
     SetBestHeight(connOptions.nBestHeight);
 
+    clientInterface = connOptions.uiInterface;
+
+    if (!InitBinds(connOptions.vBinds)) {
+        strError = _("Failed to listen on any port. Use -listen=0 if you want this.");
```
Comment (theuni) [inline src/net.cpp]: This strError param is clunky (my fault!). We should do away with it and just use clientInterface->InitError() directly. That can be done as a follow-up though.

File: src/init.cpp [hunk 8]
```diff
@@ -898,10 +879,16 @@ bool AppInitParameterInteraction()
             return InitError(_("Prune mode is incompatible with -txindex."));
     }
 
+    // -bind and -whitebind can't be set when not listening
+    size_t nUserBind =
+        (gArgs.IsArgSet("-bind") ? gArgs.GetArgs("-bind").size() : 0) +
+        (gArgs.IsArgSet("-whitebind") ? gArgs.GetArgs("-whitebind").size() : 0);
+    if (nUserBind != 0 && !GetBoolArg("-listen", DEFAULT_LISTEN)) {
```
Comment (theuni) [inline src/init.cpp]: s/GetBoolArg/gArgs.GetBoolArg/

**Comment by theuni** [comment]
looks good aside from those few issues.

File: src/net.cpp (hunk 6 above)
Comment (benma) [inline src/net.cpp]: Oops, was a bit too eager here. Re-added the `fListen` check as you suggested.

File: src/net.cpp (hunk 7 above)
Comment (benma) [inline src/net.cpp]: Done.

File: src/init.cpp (hunk 8 above)
Comment (benma) [inline src/init.cpp]: GetBoolArg() is simply wrapping gArgs.GetBoolArg(), and both variants are already used in the function. Could make that consistent in a separate PR.

**Comment by benma** [comment]
@theuni thanks for the feedback. I added a new commit for easier review, I will squash in the end.

File: src/net.cpp (hunk 4 above)
Comment (benma) [inline src/net.cpp]: > For example, there's no way to specify "listen on this address as well as the localhost fallbacks".

But that wasn't possible before either.

Do you prefer the commit which passes the whitelist flag in the pair (https://github.com/bitcoin/bitcoin/pull/10496/commits/768c7f72d9fac542a45ad788713faa1a3f46897c), or should I drop it? I don't mind either way, but the new code seems less readable than the code before, but it is semantically exactly the same.

File: src/init.cpp (hunk 8 above)
Comment (theuni) [inline src/init.cpp]: gArgs is the new (as of very recently) way of handling these args, it'd be helpful if new code didn't reintroduce the former usage as I assume the wrappers are intended to be temporary.

File: src/net.cpp (hunk 4 above)
Comment (theuni) [inline src/net.cpp]: > But that wasn't possible before either.

Sure, but we've been slowly moving these config options out to the caller so that less is hard-coded.

In the near-future, I'm looking forward to creating 2 CConnman instances and running them against each-other for testing. The more we can control via setup options, the better. I have no problem with adding those options incrementally, I'd just like to make sure we're not making future changes tougher.

As for the whitelist flag, either way is fine for this PR. Whichever you think is more readable. We'll expand on the options when needed.

Comment (benma) [inline src/net.cpp]: :+1: I reverted it as I find the previous version more readable.

**Comment by benma** [comment]
@theuni addressed the GetBoolArgs issue and squashed.

**Comment by benma** [comment]
Rebased.

**Review by theuni — APPROVED** [review]

**Comment by theuni** [comment]
utACK 07b2afef10bb6366a270e325fd41a8bc526c9ef3

**Review by ryanofsky — COMMENTED** [review]
utACK 07b2afef10bb6366a270e325fd41a8bc526c9ef3. There were many changes since last review, all discussed above.

**Comment by laanwj** [comment]
utACK 07b2afe
